### PR TITLE
User proper write-once semantics for GCS repository

### PR DESCRIPTION
**Pull request by ywelsch**
There's no need for an extra blobExists() call when writing a blob to the GCS service. GCS provides an option (with stronger consistency guarantees) on the insert method that guarantees that the blob that's uploaded does not already exist.

I have not tested this against the real GCS service, only used the API docs to develop this. @tlrx can you help me check that this works against the actual GCS service? Thanks.

Relates to #19749

### CODE REVIEW

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java [hunk 1]
```diff
@@ -209,8 +211,16 @@ void writeBlob(String blobName, InputStream inputStream, long blobSize) throws I
...
             insert.setName(blobName);
-            insert.execute();
+            try {
+                insert.execute();
+            } catch (GoogleJsonResponseException e) {
+                GoogleJsonError error = e.getDetails();
+                if ((e.getStatusCode() == HTTP_PRECON_FAILED) || ((error != null) && (error.getCode() == HTTP_PRECON_FAILED))) {
+                    throw new FileAlreadyExistsException(blobName, null, e.getMessage());
+                }
+            }
```
Comment (tlrx) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java]: Exception must be thrown again

**Review by tlrx — APPROVED** [review]
LGTM - Left one important comment though

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java (hunk 1 above)
Comment (ywelsch) [reply plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java]: right, thanks

**Comment by ywelsch** [comment]
@tlrx I've had to completely rewrite this PR due to #30168. Can you give it another full review please?

**Review by tlrx — APPROVED** [review]
Tested and approved, thanks @ywelsch

**Comment by ywelsch** [comment]
Thanks @tlrx

**Pull request merged**